docs/language.md

```diff
@@ -0,0 +1,87 @@
+# Beans Language
+
+A beans program is made up of a series of functions, the order of these functions does not matter and they can be references before being called.
+
+## Basics
+
+* A symbol is an ASCII string starting with a letter from A-Z and then followed by letters, numbers, or an underscore.
+
+## Types
+
+There are several builtin types. 
+* void 		 - represents a lack of any value
+* i8/i16/i32/i64 - signed integer types with the specified number of bits
+* u8/u16/u32/u64 - unsigned integer types with the specified number of bits
+* bool 		 - either ``true`` or ``false``
+
+User defined types are not implemented, and the design is still changing.
+
+## Functions
+
+A function has several parts.
+* A name - a symbol
+* A list of parameters wrapped in parenthesis where a parameter is a symbol followed by a type
+* An optional return type after the parameters, if omitted the return type is void
+* A block of statements wrapped in curly braces.
+
+### Examples (actual code omitted): 
+
+```
+add2(v1 i32, v2 i32) i32 {}
+```
+
+```
+main() {}
+```
+
+## Statements
+
+A statement makes up functions, but beans has most language features as expressions so they can be composed in powerful ways. However there are still a couple features that are only represented as statements.
+
+* Let        - introduces a new name into the current local scope, optionally giving it an initial value
+* Return     - returns from the current function, optionally with a value
+* Expression - just an expression which is computed, and the result is discarded
```

Comment:
Are expressions that can be proved to have no side effects valid?

Reply:
Currently yes, and I don't see a reason to disallow pure expressions specifically, though a case can be made for not allowing expressions that evaluate to a non-unit value.